### PR TITLE
Make .then() return a real Promise

### LINKED ISSUE #722: The .then() function is not compatible with Promises
The faux `.then()` function added in 1.3 completely misses the point of promises and breaks even in simple cases.

```js
request.get(someCorrectUrl)
    .then(function(){return 1;})
    .then(console.log.bind(console));
```
Should print `1`, but throws an unhandled exception.

```js
request.get('fail').then(function(){});
```
> TypeError: reject is not a function

```js
request.get('fail')
    .then(undefined, function(){throw Error();})
    .catch(function(){})
```
> .then(...).catch is not a function

Please don't try to reimplement promises — just enable use of a correct library.

I suggest:

```js
Request.prototype.then = function(resolve, reject) {
  var self = this;
  return new Promise(function(resolve, reject){
    self.end(function(err, res){
      if (err) reject(err); else resolve(res);
    });
  })
  .then(resolve, reject);
};
```

The code above assumes that promises are either supported natively (already the case in majority of browsers and even Node 0.12) or polyfilled by the user (which is a common practice, e.g. ES6 transpilers do it).

### LINKED ISSUE #852: Fake Promise is a breaking API change
The fake promise that was introduced in #709 is a **BREAKING** change and not backwards compatible.

I wrote a little scenario involving an async action via a promise. Before this change was made, this was perfectly valid until the fake promise was added to the library. For this scenario, I used native promises but have also tested with a promise library with the same result.

```js
var http = require('http')
var fs = require('fs')
var superagent = require('superagent')
var dest = fs.createWriteStream('response.txt')

// stand up a simple http server
var server = http.createServer(function(req, res) {
  console.log('got request')
  res.end('hello superagent!')
})

// Start listening
server.listen(1234, function() {

  console.log('server listening')
  
  // Generate dummy data
  var promise = new Promise(function(resolve, reject) {
    
    resolve({ hello: 'superagent' })
    
  }).then(function(data) {
    
    // generate a dummy request
    return superagent.post('http://localhost:1234').send(data)
    
  }).then(function(res) {
    // At this point, we should have a request
   // that hasn't been fired yet, based on the docs.

    // Pipe to the dest, fires the request and streams to the file
    res.pipe(dest)
    // kill the program
    dest.once('close', function() {
      console.log('wrote response')
      process.exit()
    })
    
  })
  
})
```

With `superagent@<1.3.0`, the code works as expected and you will see a `wrote response` and the `response.txt` will have the correct text `hello superagent!`. 

The problem with `superagent@^1.3.0` the request ( that should not have fired based on the docs ) is seen as a "thenable" and so the promise chain will call the `then()` method which fires off the request and returns the buffered response to the next promise chain call. Because of this, the pipe method will not work because the stream has already been exhausted, the program will never end, and the `response.txt` will have no content.

### PR DESCRIPTION
**Pull request by pornel**
Another take on Promises. 

* Uses global `Promise`. ES6 is final now, so I think it's OK to expect it for interoperability. 
   Node 0.10—the IE6 of Nodes—can easily polyfill it.
* Re-returns the same result, so that usage of the request as thenable in multiple places won't make multiple calls. That's consistent with how real promises behave.

Fixes #722 Closes #852
 
#854 #726

### CODE REVIEW

File: test/basic.js [hunk 1]
```diff
@@ -64,6 +64,36 @@ describe('request', function(){
         done();
       });
     })
+
+    it('is optional with a promise', function() {
```
Comment (focusaurus) [inline test/basic.js]: I think you need to take a `done()` parameter here or mocha will consider the test passed before all the `assert` calls are reached.

**Comment by focusaurus** [comment]
I don't see `.then()` mentioned anywhere in `docs/index.md`. Let's document this while we're at it. Other than docs and my comment above about `done()` this looks good to me.

File: test/basic.js (hunk 1 above)
Comment (pornel) [inline test/basic.js]: I'm using Mocha's Promise support here: https://mochajs.org/#working-with-promises

Comment (focusaurus) [inline test/basic.js]: Ah, I haven't seen that before. Looks good then.

**Comment by ewnd9** [comment]
Is it possible to access a response body in the catch method (e.g. server is responding with a 500 status code, but an error reason is in a body)?

**Comment by pornel** [comment]
I many cases we're adding properties to the error object. Check for `err.response`.

**Pull request merged**